src/natcap/invest/seasonal_water_yield/seasonal_water_yield.py

```diff
@@ -879,133 +875,178 @@ def _calculate_annual_qfi(qfm_path_list, target_qf_path):
-        # on large valid_si/a_im this number will be zero and the latter
-        # exponent will also be zero because of a divide by zero. rather than
-        # raise that numerical warning, just handle it manually
-        E1 = scipy.special.expn(1, valid_si / a_im)
-        E1[valid_si == 0] = 0
-        nonzero_e1_mask = E1 != 0
-        exp_result = numpy.zeros(valid_si.shape)
-        exp_result[nonzero_e1_mask] = numpy.exp(
-            (0.8 * valid_si[nonzero_e1_mask]) / a_im[nonzero_e1_mask] +
-            numpy.log(E1[nonzero_e1_mask]))
-
-        # qf_im is the quickflow at pixel i on month m Eq. [1]
-        qf_im[valid_mask] = (25.4 * valid_n_events * (
-            (a_im - valid_si) * numpy.exp(-0.2 * valid_si / a_im) +
-            valid_si ** 2 / a_im * exp_result))
-
-        # if precip is 0, then QF should be zero
-        qf_im[(p_im == 0) | (n_events == 0)] = 0.0
-        # if we're on a stream, set quickflow to the precipitation
-        valid_stream_precip_mask = stream_array == 1
-        if p_nodata is not None:
-            valid_stream_precip_mask &= ~utils.array_equals_nodata(
-                p_im, p_nodata)
-        qf_im[valid_stream_precip_mask] = p_im[valid_stream_precip_mask]
+        a_im = p_im[case_3_mask] / (n_m[case_3_mask] * 25.4)
+
+        # don't raise a warning on overflow
+        with numpy.errstate(over='ignore'):
+            exp_result = numpy.exp((0.8 * s_i[case_3_mask]) / a_im)
+
+        # edge case 3a: set the whole term to 0 when s_i = 0
+        subcase_a_mask = s_i[case_3_mask] == 0
+        term_result[subcase_a_mask] = 0
+
+        # edge case 3b: set the whole term to 0 when exp() overflows
+        subcase_b_mask = numpy.isinf(exp_result)
+        term_result[subcase_b_mask] = 0
+
+        # case 3c: evaluate the term as usual
+        # s_i^2 / a_im * exp(0.8 * s_i / a_im) * E1(s_i / a_im)
+        subcase_c_mask = ~(subcase_a_mask | subcase_b_mask)
+        term_result[subcase_c_mask] = (
+            s_i[case_3_mask][subcase_c_mask] ** 2 / a_im[subcase_c_mask] *
+            exp_result[subcase_c_mask] *
+            scipy.special.exp1(
+                s_i[case_3_mask][subcase_c_mask] / a_im[subcase_c_mask]
+            )
+        )
+
+        qf_im[case_3_mask] = (
+            25.4 * n_m[case_3_mask] * (
+                (a_im - s_i[case_3_mask]) *
+                numpy.exp(-0.2 * s_i[case_3_mask] / a_im) +
+                term_result
+            )
+        )
 
         # this handles some user cases where they don't have data defined on
         # their landcover raster. It otherwise crashes later with some NaNs.
         # more intermediate outputs with nodata values guaranteed to be defined
-        qf_im[utils.array_equals_nodata(qf_im, qf_nodata) &
-              ~utils.array_equals_nodata(stream_array, stream_nodata)] = 0.0
+        qf_im[utils.array_equals_nodata(qf_im, TARGET_NODATA) &
+              ~utils.array_equals_nodata(stream, stream_nodata)] = 0
```

Comment:
This part seems suspicious, but I'm going to make another issue about it.

Reply:
This is odd. This is setting nodata values to 0 where `qf_im == Nodata & stream != Nodata`... This would indicate only Stream nodata values would persist through as nodata. Huh.